helpers/target.go

```diff
@@ -129,3 +154,332 @@ func isAllowed(addr string) (bool, error) {
 	}
 	return true, nil
 }
+
+// ServiceCreds represents the credentials
+// necessary to access an authenticated service.
+// There are constructors available in this same
+// package for:
+//    - AWS Assume role through vulcan-assume-role svc.
+//    - Docker registry.
+//    - Github repository.
+type ServiceCreds interface {
+	URL() string
+	Username() string
+	Password() string
+}
+
+// AWSCreds holds data required
+// to perform an assume role request.
+type AWSCreds struct {
+	AssumeRoleURL string
+	Role          string
+}
+
+// NewAWSCreds creates a new AWS Credentials for Assume Role.
+func NewAWSCreds(assumeRoleURL, role string) *AWSCreds {
+	return &AWSCreds{
+		AssumeRoleURL: assumeRoleURL,
+		Role:          role,
+	}
+}
+func (c *AWSCreds) URL() string {
+	return c.AssumeRoleURL
+}
+func (c *AWSCreds) Username() string {
+	return c.Role
+}
+func (c *AWSCreds) Password() string {
+	return ""
+}
+
+type DockerCreds struct {
+	RegistryURL string
+	User        string
+	Pass        string
+}
+
+// DockerHubCreds represents a void
+// DockerCreds struct allowed to be
+// used with Docker Hub registry.
+var DockerHubCreds = &DockerCreds{}
+
+// NewDockerCreds creates a new Docker Credentials struct.
+func NewDockerCreds(registryURL, user, pass string) *DockerCreds {
+	return &DockerCreds{
+		RegistryURL: registryURL,
+		User:        user,
+		Pass:        pass,
+	}
+}
+func (c *DockerCreds) URL() string {
+	return c.RegistryURL
+}
+func (c *DockerCreds) Username() string {
+	return c.User
+}
+func (c *DockerCreds) Password() string {
+	return c.Pass
+}
+
+type GitCreds struct {
+	RepoURL string
+	User    string
+	Pass    string
+}
+
+// NewGitCreds creates a new Git Credentials struct.
+// user and pass can be void if no auth is required.
+func NewGitCreds(repoURL, user, pass string) *GitCreds {
+	return &GitCreds{
+		RepoURL: repoURL,
+		User:    user,
+		Pass:    pass,
+	}
+}
+func (c *GitCreds) URL() string {
+	return c.RepoURL
+}
+func (c *GitCreds) Username() string {
+	return c.User
+}
+func (c *GitCreds) Password() string {
+	return c.Pass
+}
+
+// IsReachable returns wether target is reachable
+// so the check execution can be performed.
+// If asset type is void, it will be inferred from target.
+//
+// ServiceCredentials are required for AWS, Docker and Git types.
+// Constructors for AWS, Docker and Git credentials can be found
+// in this same package.
+//
+// Verifications made depend on the asset type:
+//    - IP: None.
+//    - IPRange: None.
+//    - DomainName: None.
+//    - Hostname: NS Lookup resolution.
+//    - WebAddress: HTTP GET request.
+//    - AWSAccount: Assume Role.
+//    - DockerImage: Docker pull.
+//    - GitRepository: Git clone.
```

Comment:
I have no idea but there are any other actions to perform lighter operations that would confirm the access? It's true what you told me about caching and so, but asking just in case.

Reply:
Actually I tried to do that initially, particularly for these two cases of course. I tried to find a way to verify if a Docker image existed in a registry, but did not find an easy way to do it with our DockerTools library, or with the DockerClient library itself, that's why I opted to do a pull, because in any case that will most probably done by the check itself also.

But I can try to dig deeper on this to see if there is some alternative.
The same reasoning applied for the Git repo, but I can check again too.

Reply:
UPDATE: I modified the Git verification to perform a `ls-remote` instead of clone.
I reviewed the Docker one, but does not see so easy, also there are two possible versions for the API.
Also, this makes me think that maybe then we should not export these "individual methods" and discard any possible output like the AWS STS creds in the case of the assume role.